src/SFML/Window/Android/WindowImplAndroid.cpp

```diff
@@ -716,11 +717,12 @@ char32_t WindowImplAndroid::getUnicode(AInputEvent* event)
     const std::lock_guard lock(states.mutex);
 
     // Prepare the Java virtual machine
-    JNIEnv* lJNIEnv = states.activity->env;
-    auto    jni     = Jni::attachCurrentThread(states.activity->vm, &lJNIEnv);
+    auto jni = Jni::attachCurrentThread(*states.activity);
     if (!jni)
         err() << "Failed to initialize JNI, couldn't get the Unicode value" << std::endl;
 
+    auto& lJNIEnv = jni->getEnv();
```

Comment:
What was the reasoning for adding `getEnv`? It appears that anywhere we use `getEnv` we could also access the environment through `states.activity`.

Reply:
I thought I'll give the Jni object a little more reason to live. You know, as you've been simplifying those if conditions, if env was supplied by Jni, you wouldn't have the reason to do so in the first place. Also, env is only valid for as long as the Jni object itself, so it makes more sense.